pricegraph/src/graph/bellman_ford.rs

```diff
@@ -2,40 +2,170 @@
 //! implementation of the Bellman-Ford graph search algorigthm that returns the
 //! detected negative cycle on error.
 
+use super::path::{find_cycle, NegativeCycle, Path};
 use petgraph::algo::FloatMeasure;
 use petgraph::visit::{
     Data, EdgeRef, GraphBase, IntoEdges, IntoNodeIdentifiers, NodeCount, NodeIndexable,
 };
 
-/// A type definition for the result of a Bellman-Ford shortest path search
-/// containing the weight of the shortest paths and the predecessor vector for
-/// the shortests paths.
-pub type Paths<G> = (
-    Vec<<G as Data>::EdgeWeight>,
-    Vec<Option<<G as GraphBase>::NodeId>>,
-);
-
-/// A negative cycle error with the node on which it was detected along with the
-/// predecessor vector that can be used to re-create the cycle path.
-#[derive(Debug)]
-pub struct NegativeCycle<N>(pub Vec<Option<N>>, pub N);
-
-/// This implementation is taken from the `petgraph` crate with a small
-/// modification to return the path when a negative cycle is detected.
+/// A vector associating to each node index a distance from the source.
+type Distance<G> = Vec<<G as Data>::EdgeWeight>;
+
+/// Stores the information needed to manage the predecessor list.
+/// For each node index, this type contains its predecessor node and
+/// distance in the graph from a source node.
+struct PredecessorStore<G: GraphBase + Data>(Vec<Option<G::NodeId>>, Distance<G>);
+
+impl<G: GraphBase + Data> PredecessorStore<G> {
+    fn distance(&self, node_index: usize) -> &G::EdgeWeight {
+        let PredecessorStore(_, distance) = self;
```

Comment:
Seeing that you always destructure `self` back into a tuple and give the fields names, it makes more sense to have `PredecessorStore` be a struct with named fields in the first place like

```rust
struct PredecessorStore<G: GraphBase + Data> {
    predecessor: Vec<Option<G::NodeId>>,
    distance: Distance<G>,
}
```
Or is there another reason for the tuple struct?
You can also access the fields without naming them first btw:

```rust
&self.1[node_index]
```

Reply:
True, I'll change it to a struct.
I wrote it this way because I plan to make `PredecessorStore` into an `enum`, so with this destructuring the diff would be smaller when the change happen. On the other hand, it's better to use a `struct` with named fields here to keep this PR cleaner.

Reply:
An enum can be struct-like with named fields too btw, if you prefer that.